### ISSUE TITLE
Towards a pluggable Reds

### ISSUE THREAD

**Issue by stockholmux** [tj/reds#45]
This module, in my mind, is at stable state. It's quick, it works well generally and has a good number of depending modules (especially since it is used in Kue).

There is a lot that can be done with the basic structure and it can be monkey patched for extended functionality, but this comes with baggage and is not ideal. There have been a few good PRs that have requested extended and tweaked functionality but perhaps those PRs wouldn't make sense for all. I'm proposing keeping the functionality the same but making `reds` more extensible and pluggable. I'd like to kick this process off by opening the floor to ideas on how to accomplish this in the most flexible and performant way.

My initial thoughts on the areas of plugablity:
- index / remove
- natural language functions (stem / metaphone / stop words, etc.)
- search query

The plugins could be introduced when creating the new search with an optional config object. It might look something like this:

```
var search = reds.createSearch('my-search', { indexRemovePlugin : myIndexRemovePlugin })
```

Thoughts on this idea? Have I missed something?

**Comment by stockholmux** [comment]
[Pluggable version](https://github.com/stockholmux/reds) and a [general writeup](https://medium.com/@stockholmux/reds-internals-searching-and-better-searching-with-node-js-and-redis-57da99077e83).

**Comment by plus-** [comment]
Hey, cool job.

Not sure why you are exposing the default functions in the export though.

**Comment by stockholmux** [comment]
Resolved with #50
